pkg/obfuscate/shell.go

```diff
@@ -0,0 +1,607 @@
+
+			// Call this function recursively
+			obfuscatedSlices = obfuscateShellCommandToken(subcommandTokens)
+
+			// Increment the index to the end of the subcommand
+			index = endSubcommandToken
+			break
+		}
+	}
+
+	return index, obfuscatedSlices
+}
+
+// grabFullArgument grabs the full argument from the given tokens starting at the given index.
+// returns the index of the last token of the argument and the number of tokens grabbed.
+// For examples:
+// - if the argument starting at index 3 is a Dollar token, then it should also grab the next token (calling itself recursively) as part of the argument.
+// - if the argument starting at index 3 is a Field token, then it should only grab the Field token.
+// - if the argument starting at index 3 is an Equal token, then it should grab the Equal token and the next token (calling itself recursively).
+// - if the argument starting at index 3 is a SingleQuote or DoubleQuote token, then it should grab the whole quoted string.
+func grabFullArgument(tokens []ShellToken, index int) (int, int) {
+	tokensLength := len(tokens)
+	if index >= tokensLength {
+		return tokensLength - 1, 0 // Return the last token index if we are out of bounds
+	}
+
+	nbrGrabbed := 0
+	for ; index < tokensLength; index++ {
+		kind := tokens[index].Kind
+
+		// We can't grab a control or redirection token
+		if kind == Control || kind == Redirection {
+			return index - 1, nbrGrabbed
+		}
+
+		// Grab only the current token
+		if kind == Field || kind == Equal || kind == ShellVariable {
+			return index, nbrGrabbed + 1
+		}
+
+		// Grab the next token
+		if kind == Dollar || kind == Equal {
+			nbrGrabbed++
+			// continue
+		} else if kind == ParentheseOpen {
+			// Grab the whole parentheses content
+			j := index + 1
+			for ; j < tokensLength && tokens[j].Kind != ParentheseClose; j++ {
+				if tokens[j].Kind == Dollar {
+					// Grab the next token
+					nbrGrabbed++
+					// continue
+				} else if tokens[j].Kind == ParentheseOpen {
+					// Recursively grab the whole parentheses content
+					newIndex, newNbrGrabbed := grabFullArgument(tokens, j)
+					nbrGrabbed += newNbrGrabbed
+					j = newIndex
+				} else {
+					nbrGrabbed++
+				}
+			}
+
+			index = j
+
+			// Add the last token if it is the same as the start token
+			if index < tokensLength && tokens[index].Kind == ParentheseClose {
+				nbrGrabbed++
+			}
+
+			break
+		} else if kind == SingleQuote || kind == DoubleQuote || kind == Backticks {
+			nbrGrabbed++
+			index++
+
+			// Grab the whole quoted string
+			for ; index < tokensLength && tokens[index].Kind != kind; index++ {
+				nbrGrabbed++
+			}
+
+			// Add the last token if it is the same as the start token
+			if index < tokensLength && tokens[index].Kind == kind {
+				nbrGrabbed++
+			}
+
+			break
+		} else if kind == Executable {
+			nbrGrabbed++
+			break
+		} else {
+			// Not a known token kind
```

Comment:
should we log a warning or debug or something here to make sure this doesn't happen?